src/Lecture4.hs

```diff
@@ -131,8 +140,28 @@ errors. We will simply return an optional result here.
 🕯 HINT: Use the 'readMaybe' function from the 'Text.Read' module.
 -}
 
+parseMaybe :: ReadP a -> String -> Maybe a
+parseMaybe parser input =
+    case readP_to_S parser input of
+        [] -> Nothing
+        ((result, _):_) -> Just result
+
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow = parseMaybe rowParser
+
+rowParser :: ReadP Row
+rowParser = do
+    productName <- many1 get
+    _ <- satisfy (== ',')
+    skipSpaces
+    trade <- fmap read (string "Buy" <|> string "Sell")
+    skipSpaces
+    _ <- satisfy (== ',')
+    skipSpaces
+    cost <- fmap read (many1 (satisfy isDigit))
+    skipSpaces
+    eof
+    pure (Row productName trade cost)
```

Comment:
I've read [Parse, don't validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/), so decided to go with parser, since splitting/droping/matching I already can, and with parsers there was a learning opportunity. I've seen that people use `attoparsec`, `megaparsec`, `trifecta`, and nobody uses ReadP. It's just I wanted to not to bring external packages and only use things included in base.

Reply:
Practice is good 🙂  The `ReadP` parser is used in derived instances of the `Read` typeclass. But usually, people don't write such parsers manually. Using external libraries like the ones you mentioned is a better choice usually 👍🏻 

The `parseRow` function already implements "parse, don't validate" concept. The functions takes some unstructured representation (`String`) and returns a value `Maybe Row` instead of something like `isRow :: String -> Bool` and `toRow :: String -> Row`.

For this exercise using `ReadP` is fine and it's a good practice on parser combinators 🙂